internal/controllers/gateway/gateway_utils.go

```diff
@@ -670,3 +666,71 @@ func isTLSSecretValid(secret *corev1.Secret) bool {
 	}
 	return true
 }
+
+func routeAcceptedByGateways(routeNamespace string, parentStatuses []RouteParentStatus) []k8stypes.NamespacedName {
+	gateways := []k8stypes.NamespacedName{}
+	for _, routeParentStatus := range parentStatuses {
+		gatewayNamespace := routeNamespace
+		parentRef := routeParentStatus.ParentRef
+		if (parentRef.Group != nil && *parentRef.Group != gatewayV1beta1Group) ||
+			(parentRef.Kind != nil && *parentRef.Kind != "Gateway") {
+			continue
+		}
+		if parentRef.Namespace != nil {
+			gatewayNamespace = string(*parentRef.Namespace)
+		}
+		if lo.ContainsBy(routeParentStatus.Conditions, func(condition metav1.Condition) bool {
+			return condition.Type == string(gatewayv1beta1.RouteConditionAccepted) &&
+				condition.Status == metav1.ConditionTrue
+		}) {
+			gateways = append(gateways,
+				k8stypes.NamespacedName{
+					Namespace: gatewayNamespace,
+					Name:      string(parentRef.Name),
+				})
+		}
+	}
+	return gateways
+}
+
+// Get all the HTTPRoutes and update the listeners' attachedRoutes field.
+// At this point we take into account HTTPRoutes only, as they are the
+// only routes in beta.
+func getAttachedRoutesForListener(ctx context.Context, client client.Client, listener gatewayv1beta1.Listener, gateway gatewayv1beta1.Gateway) (int32, error) {
+	httpRouteList := gatewayv1beta1.HTTPRouteList{}
+	if err := client.List(ctx, &httpRouteList); err != nil {
+		return 0, err
+	}
+	var attachedRoutes int32
+	for _, route := range httpRouteList.Items {
```

Comment:
Need to take into account paging via `httpRouteList.Continue` token.

Similarly as it's done e.g. for EndpointsSlice https://github.com/Kong/kubernetes-ingress-controller/blob/523abce0730cfdfc6499536d6dedab06a52234d1/internal/adminapi/endpoints.go#L61-L64